README.md
## Costing Table field for Advanced Custom Fields

This [WordPress](http://www.wordpress.org/) plugin extends [Advanced Custom Fields](http://www.advancedcustomfields.com) 5.x+ with a new field called "Costing Table".

Used by the author's for personal projects and company projects in the travel industry that require tabular data calculations for determining the cost of travel packages products, similar to that of a spreadsheet, but inside WordPress and using Advanced Custom Fields framework.
Field data is stored as an object and redistributed across the backend table cells via javascript. By retrieving the field using `get_field( $field_name, $post_id )` function which ships witch ACF, it will return an object with table calculations and totals.

To install copy to WordPress plugins directory and then activate from the plugins dashboard inside your WordPress admin area. Advanced Custom Fields needs to be installed and activated first.

No support or documentation is offered for this plugin. It is not designed to work with Advanced Custom Fields version 3.x or 4.x.


ChangeLog.md

### 1.0.0 (08 Jul 2014)
* First public release


acf-costing-table.php
<?php
/**
 * Advanced Custom Fields - Costing Table field
 *
 * A WordPress plugin to extend Advanced Custom Fields 5 with a new field type meant for tabular costing data calculations.
 *
 * @package   ACF Costing Table
 * @author    nekojira <fulvio@nekojira.com>
 * @license   GPL-2.0+
 * @link      https://github.com/nekojira/acf-costing-table
 * @copyright 2014 nekojira
 *
 * @wordpress-plugin
 * Plugin Name:       Advanced Custom Fields: Costing Table field
 * Plugin URI:        https://github.com/nekojira/acf-costing-table
 * Description:       Costing Table field for Advanced Custom Fields 5.
 * Version:           1.0.0
 * Author:            nekojira
 * Author URI:        https://github.com/nekojira/
 * Text Domain:       acf-costing_table
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * Domain Path:       /lang
 * GitHub Plugin URI: https://github.com/nekojira/acf-costing-table
 */

// abort if file is called directly
if ( ! defined( 'WPINC' ) )
	die;

// set plugin text domain
load_plugin_textdomain( 'acf-costing-table', false, dirname( plugin_basename(__FILE__) ) . '/lang/' );

// plug into ACF5
add_action('acf/include_field_types', function() { require_once 'field/class-acf-costing-table.php'; } );

field/class-acf-costing-table.php
<?php
/**
 * ACF Costing Table field
 *
 * @package   ACF Costing Table
 * @author    nekojira <fulvio@nekojira.com>
 * @license   GPL-2.0+
 * @link      https://github.com/nekojira/acf-costing-table/
 * @copyright 2014 nekojira
 */

/**
 * Class ACF Costing Table
 * Creates a costing table ACF field
 *
 * @since 1.0.0
 */
class ACF_Costing_Table extends acf_field {

	/**
	 * Setup field data
	 *
	 * @since 1.0.0
	 */
	function __construct() {

		// field name
		$this->name = 'costing_table';
		$this->label = __( 'Costing table', 'acf-costing_table' );
		// field type
		$this->category = 'jquery';
		// default settings
		$this->defaults = array();

		// do not delete!
		parent::__construct();

	}

	/**
	 * Field settings
	 * Creates extra settings for the field, when editing a new field
	 *
	 * @param	array	$field	the field
	 *
	 * @since	1.0.0
	 *
	 * @return	void
	 */
	function render_field_settings( $field ) {

		// currently no options are offered

	}

	/**
	 * Render the field HTML
	 * Creates the HTML interface of the field
	 *
	 * @param	array	$field	the field
	 *
	 * @since 	1.0.0
	 *
	 * @return void
	 */
	function render_field( $field ) {

		require_once 'includes/acf-costing-table-html.php';

	}

	/**
	 * Add CSS and Javascripts when render_field() is called
	 * This action is called in the admin_enqueue_scripts action on the edit screen where the field is created.
	 *
	 * @since	1.0.0
	 *
	 * @return void
	 */
	function input_admin_enqueue_scripts() {

		$dir = plugin_dir_url( __FILE__ );

		// register & include JS
		wp_register_script( 'acf-input-costing_table', "{$dir}assets/js/min/input.min.js" );
		wp_enqueue_script('acf-input-costing_table');

		// register & include CSS
		wp_register_style( 'acf-input-costing_table', "{$dir}assets/css/min/input.css" );
		wp_enqueue_style('acf-input-costing_table');

	}

	/**
	 * Format value
	 *
	 * This filter is applied to the $value after it is loaded from the db and before it is returned to the template
	 *
	 * @param	mixed	$value		the field value
	 * @param 	int		$post_id	post to which the field is attached
	 * @param 	object	$field		the field
	 *
	 * @since	1.0.0
	 *
	 * @return	array	will return an array with the costing table totals
	 */
	function format_value( $value, $post_id, $field ) {

		// bail early if no value
		if( empty($value) ) {

			return $value;

		}

		$data =  json_decode( $value );

		return $data;
	}

}

// create field
new ACF_Costing_Table();

field/includes/acf-costing-table-html.php
<?php if ( isset( $field ) ) : ?>

	<div id="<?php echo $field['id'] ?>" class="costing_table">

		<?php

		// prepare saved data
		$data = (array) json_decode( $field['value'] );
		$days = isset( $data['per_day'] ) ? $data['per_day'] : '';
		$totals = isset( $data['finals'] ) ? $data['finals'] : '';

		?>

		<div class="acf-costing_table-buttons">
			<ul>
				<li><a class="acf-costing_table-add_day acf-button button" title="<?php _e( 'Add day', 'acf-costing_table' ); ?>">+</a>
				</li>
				<li>
					<a class="acf-costing_table-remove_day acf-button button" title="<?php _e( 'Remove last row', 'acf-costing_table' ); ?>">&ndash;</a>
				</li>
				<li>
					<a class="acf-costing_table-calc acf-button button button-primary" title="<?php _e( 'Warning! Any existing values will be overwritten!', 'acf-costing_table' ); ?>"><?php _e( 'Calculate', 'acf-costing_table' ); ?></a>
				</li>
			</ul>
		</div>

		<table class="acf-costing_table widefat acf-input-table">

			<thead>
				<tr>
					<th><small>#</small></th>
					<th colspan="2"><small><?php _e( 'Fixed costs', 'acf-costing_table' ); ?></small></th>
					<th colspan="9"><small><?php _e( 'Transportation costs per group of people', 'acf-costing_table' ); ?></small></th>
					<th colspan="3"><small><?php _e( 'Costs per individual', 'acf-costing_table' ); ?></small></th>
				</tr>
				<tr>
					<th data-field="day"><?php _e( 'Day', 'acf-costing_table' ); ?></th>
					<th data-field="base"><?php _e( 'Guides', 'acf-costing_table' ); ?></th>
					<th data-field="staff"><?php _e( 'Staff', 'acf-costing_table' ); ?></th>
					<th data-field="1">1</th>
					<th data-field="2">2</th>
					<th data-field="3">3</th>
					<th data-field="4">4</th>
					<th data-field="5">5-6</th>
					<th data-field="7">7-8</th>
					<th data-field="9">9-12</th>
					<th data-field="13">13-24</th>
					<th data-field="25">25-39</th>
					<th data-field="meals"><?php _e( 'Meals', 'acf-costing_table' ); ?></th>
					<th data-field="tickets"><?php _e( 'Entrances', 'acf-costing_table' ); ?></th>
					<th data-field="misc"><?php _e( 'Misc', 'acf-costing_table' ); ?></th>
				</tr>
			</thead>

			<tfoot>
				<tr>
					<th colspan="15"><small><?php _e( 'Totals', 'acf-costing_table' ) ?></small></th>
				</tr>
				<tr>
					<th data-field="per_person" colspan="3"><?php _e( 'Per person', 'acf-costing_table' ); ?></th>
					<th data-field="1">1</th>
					<th data-field="2">2</th>
					<th data-field="3">3</th>
					<th data-field="4">4</th>
					<th data-field="5">5-6</th>
					<th data-field="7">7-8</th>
					<th data-field="9">9-12</th>
					<th data-field="13">13-16</th>
					<th data-field="17">17-20</th>
					<th data-field="21">21-24</th>
					<th data-field="25">25-34</th>
					<th data-field="35">34-39</th>
				</tr>
				<tr class="acf-costing_totals">
					<?php

					$base_var = isset( $totals->per_person ) ? $totals->per_person : '';
					$base_1 = isset( $totals->{strval('1')} ) ? $totals->{strval('1')} : 0;
					$base_2 = isset( $totals->{strval('2')} ) ? $totals->{strval('2')} : 0;
					$base_3 = isset( $totals->{strval('3')} ) ? $totals->{strval('3')} : 0;
					$base_4 = isset( $totals->{strval('4')} ) ? $totals->{strval('4')} : 0;
					$base_5 = isset( $totals->{strval('5')} ) ? $totals->{strval('5')} : 0;
					$base_7 = isset( $totals->{strval('7')} ) ? $totals->{strval('7')} : 0;
					$base_9 = isset( $totals->{strval('9')} ) ? $totals->{strval('9')} : 0;
					$base_13 = isset( $totals->{strval('13')} ) ? $totals->{strval('13')} : 0;
					$base_17 = isset( $totals->{strval('17')} ) ? $totals->{strval('17')} : 0;
					$base_21 = isset( $totals->{strval('21')} ) ? $totals->{strval('13')} : 0;
					$base_25 = isset( $totals->{strval('25')} ) ? $totals->{strval('25')} : 0;
					$base_35 = isset( $totals->{strval('35')} ) ? $totals->{strval('35')} : 0;

					?>
					<th colspan="3"><input type="number" class="base_total_var" value="<?php echo $base_var; ?>"></th>
					<th><input type="number" class="base_total_1" min="0" value="<?php echo $base_1; ?>" /></th>
					<th><input type="number" class="base_total_2" min="0" value="<?php echo $base_2; ?>" /></th>
					<th><input type="number" class="base_total_3" min="0" value="<?php echo $base_3; ?>" /></th>
					<th><input type="number" class="base_total_4" min="0" value="<?php echo $base_4; ?>" /></th>
					<th><input type="number" class="base_total_5-6" min="0" value="<?php echo $base_5; ?>" /></th>
					<th><input type="number" class="base_total_7-8" min="0" value="<?php echo $base_7; ?>" /></th>
					<th><input type="number" class="base_total_9-12" min="0" value="<?php echo $base_9; ?>" /></th>
					<th><input type="number" class="base_total_13-16" min="0" value="<?php echo $base_13; ?>" /></th>
					<th><input type="number" class="base_total_17-20" min="0" value="<?php echo $base_17; ?>" /></th>
					<th><input type="number" class="base_total_21-24" min="0" value="<?php echo $base_21; ?>" /></th>
					<th><input type="number" class="base_total_25-34" min="0" value="<?php echo $base_25; ?>" /></th>
					<th><input type="number" class="base_total_35-39" min="0" value="<?php echo $base_35; ?>" /></th>
				</tr>
			</tfoot>

			<tbody>
				<?php

				if ( is_array( $days ) ) :

					$count = 1;
					foreach ( $days as $day => $value ) :

						$row = '<tr class="costing_day costing_day_' . $count . '">';
							$row .= '<td class="day_count">' . $count . '</td>';
							$row .= '<td class="day_base">		<input type="number" class="day_' . $count . '_base" 	value="' . $value->base .'"></td>';
							$row .= '<td class="day_staff">		<input type="number" class="day_' . $count . '_staff" 	value="' . $value->staff . '"></td>';
							$row .= '<td class="day_tour_1">	<input type="number" class="day_' . $count . '_tour_1" 	value="' . $value->{strval('1')} . '"></td>';
							$row .= '<td class="day_tour_2">	<input type="number" class="day_' . $count . '_tour_2" 	value="' . $value->{strval('2')} . '"></td>';
							$row .= '<td class="day_tour_3">	<input type="number" class="day_' . $count . '_tour_2" 	value="' . $value->{strval('3')} . '"></td>';
							$row .= '<td class="day_tour_4">	<input type="number" class="day_' . $count . '_tour_2" 	value="' . $value->{strval('4')} . '"></td>';
							$row .= '<td class="day_tour_5-6">	<input type="number" class="day_' . $count . '_tour_2" 	value="' . $value->{strval('5')} . '"></td>';
							$row .= '<td class="day_tour_7-8">	<input type="number" class="day_' . $count . '_tour_2" 	value="' . $value->{strval('7')} . '"></td>';
							$row .= '<td class="day_tour_9-12">	<input type="number" class="day_' . $count . '_tour_2" 	value="' . $value->{strval('9')} . '"></td>';
							$row .= '<td class="day_tour_13-24"><input type="number" class="day_' . $count . '_tour_2" 	value="' . $value->{strval('13')} . '"></td>';
							$row .= '<td class="day_tour_25-39"><input type="number" class="day_' . $count . '_tour_2" 	value="' . $value->{strval('25')} . '"></td>';
							$row .= '<td class="day_meals">		<input type="number" class="day_' . $count . '_meals" 	value="' . $value->meals . '"></td>';
							$row .= '<td class="day_tickets">	<input type="number" class="day_' . $count . '_tickets" value="' . $value->tickets . '"></td>';
							$row .= '<td class="day_misc">		<input type="number" class="day_' . $count . '_misc" 	value="' . $value->misc . '"></td>';
						$row .= '</tr>';

						echo $row;
						$count++;

					endforeach;

				endif;

				?>
			</tbody>

		</table>

		<div class="acf-costing_table-output">
			<input type="hidden" id="<?php echo $field[ 'name' ]; ?>" name="<?php echo $field[ 'name' ]; ?>" value="<?php echo esc_attr( $field[ 'value' ] ); ?>" />
		</div>

	</div>

<?php endif; ?>

field/assets/js/input.js
(function($){

	/**
	 * Fired upon initialization of ACF field
	 *
	 * @since	1.0.0
	 *
	 * @param	jQuery selection 	$table		the jQuery element which contains the ACF field
	 */
	function initialize_field( $costing_table_field ) {

		// table elements
		var	table = $costing_table_field.find('.acf-costing_table'),
			thead = $(table).find('thead'),
			tbody = $(table).find('tbody'),
			tfoot = $(table).find('tfoot'),
			totals = $(tfoot).find('.acf-costing_totals');

		// add day row
		$('.acf-costing_table-add_day').on('click', function() {

			var row = '',
				counter = $(tbody).find('tr').length + 1;
			// row cells html
			row += '<tr class="costing_day costing_day_' + counter + '">';
				row += '<td class="day_count">' 	+ counter + '</td>';
				row += '<td class="day_base">       <input type="number" class="day_' + counter + '_base" /></td>';
				row += '<td class="day_staff">      <input type="number" class="day_' + counter + '_staff" /></td>';
				row += '<td class="day_tour_1">     <input type="number" class="day_' + counter + '_tour_1" /></td>';
				row += '<td class="day_tour_2">     <input type="number" class="day_' + counter + '_tour_2" /></td>';
				row += '<td class="day_tour_3">     <input type="number" class="day_' + counter + '_tour_3" /></td>';
				row += '<td class="day_tour_4">     <input type="number" class="day_' + counter + '_tour_4" /></td>';
				row += '<td class="day_tour_5-6">   <input type="number" class="day_' + counter + '_tour_5-6" /></td>';
				row += '<td class="day_tour_7-8">   <input type="number" class="day_' + counter + '_tour_7-8" /></td>';
				row += '<td class="day_tour_9-12">  <input type="number" class="day_' + counter + '_tour_9-12" /></td>';
				row += '<td class="day_tour_13-24"> <input type="number" class="day_' + counter + '_tour_13-24" /></td>';
				row += '<td class="day_tour_25-39"> <input type="number" class="day_' + counter + '_tour_25-39" /></td>';
				row += '<td class="day_meals">      <input type="number" class="day_' + counter + '_meals" /></td>';
				row += '<td class="day_tickets">    <input type="number" class="day_' + counter + '_tickets"  /></td>';
				row += '<td class="day_misc">       <input type="number" class="day_' + counter + '_misc" /></td>';
			row += '</tr>';

			$(tbody).append(row);

			if (counter > 1) {

				var prev = counter -1,
					prevRow = $('.costing_day_' + prev),
					newRow = $('.costing_day_' + counter),
					numCols = $(newRow).find('td').length -1;

				// copies the values of old row cells inputs into new ones appended
				for (var i = 0 ; i <= numCols ; i++) {

					newRow.find('td input').eq(i).val(prevRow.find('td input').eq(i).val());

				}

			}
			++counter;

		}); // add day

		// remove day row
		$('.acf-costing_table-remove_day').on('click', function() {

			$(tbody).find('tr:last').remove();

		}); // remove day

		/**
		 * Calculate costs
		 * Calculates costing table inputs and fills table tfoot input cells
		 *
		 * @since	1.0.0
		 */
		function ACF_costing_table_calculate() {

			var cells = $(tbody).find('tr'),
				inputs = $(cells).find('input');

			// get all the inputs in tbody rows
			inputs.each(function() {

				// assign 0 to empty or invalid inputs
				if ( $(this).val() == "" || isNaN($(this).val()) ) {
					$(this).val('0');
				}

			});

			// costing data
			var base = 0, staff = 0,
				tour_1 = 0, tour_2 = 0, tour_3 = 0, tour_4 = 0, tour_5to6 = 0, tour_7to8 = 0, tour_9to12 = 0, tour_13to24 = 0, tour_25to39 = 0,
				meals = 0, tickets = 0, misc = 0;

			$(cells).find( 'td.day_base input'			).each(function(){ base 		+= Number($(this).val()) });
			$(cells).find( 'td.day_staff input'			).each(function(){ staff 		+= Number($(this).val()) });
			$(cells).find( 'td.day_tour_1 input'		).each(function(){ tour_1 		+= Number($(this).val()) });
			$(cells).find( 'td.day_tour_2 input'		).each(function(){ tour_2 		+= Number($(this).val()) });
			$(cells).find( 'td.day_tour_3 input'		).each(function(){ tour_3 		+= Number($(this).val()) });
			$(cells).find( 'td.day_tour_4 input'		).each(function(){ tour_4 		+= Number($(this).val()) });
			$(cells).find( 'td.day_tour_5-6 input'		).each(function(){ tour_5to6 	+= Number($(this).val()) });
			$(cells).find( 'td.day_tour_7-8 input'		).each(function(){ tour_7to8 	+= Number($(this).val()) });
			$(cells).find( 'td.day_tour_9-12 input'	).each(function(){ tour_9to12 	+= Number($(this).val()) });
			$(cells).find( 'td.day_tour_13-24 input'	).each(function(){ tour_13to24 	+= Number($(this).val()) });
			$(cells).find( 'td.day_tour_25-39 input'	).each(function(){ tour_25to39 	+= Number($(this).val()) });
			$(cells).find( 'td.day_meals input'			).each(function(){ meals 		+= Number($(this).val()) });
			$(cells).find( 'td.day_tickets input'		).each(function(){ tickets 		+= Number($(this).val()) });
			$(cells).find( 'td.day_misc input'			).each(function(){ misc 		+= Number($(this).val()) });

			// fill target inputs in table footer with calculations
			$(totals).find( '.base_total_var'   ).val( parseFloat( meals + tickets + misc ) );
			$(totals).find( '.base_total_1'     ).val( parseFloat( base + staff + tour_1 ) );
			$(totals).find( '.base_total_2'     ).val( parseFloat( base + staff + tour_2 ) );
			$(totals).find( '.base_total_3'     ).val( parseFloat( base + staff + tour_3 ) );
			$(totals).find( '.base_total_4'     ).val( parseFloat( base + staff + tour_4 ) );
			$(totals).find( '.base_total_5-6'   ).val( parseFloat( base + staff + tour_5to6 ) );
			$(totals).find( '.base_total_7-8'   ).val( parseFloat( base + staff + tour_7to8 ) );
			$(totals).find( '.base_total_9-12'  ).val( parseFloat( base + staff + tour_9to12 ) );
			$(totals).find( '.base_total_13-16' ).val( parseFloat( base + staff + tour_13to24 ) );
			$(totals).find( '.base_total_17-20' ).val( parseFloat( base + staff + tour_13to24 ) );
			$(totals).find( '.base_total_21-24' ).val( parseFloat( base + staff + tour_13to24 ) );
			$(totals).find( '.base_total_25-34' ).val( parseFloat( base + staff + tour_25to39 ) );
			$(totals).find( '.base_total_35-39' ).val( parseFloat( base + staff + tour_25to39 ) );

		}

		/**
		 * Make costing table JSON
		 * Creates the JSON object to store all the field data in one single ACF value
		 *
		 * @since	1.0.0
		 */
		function ACF_costing_table_make_json() {

			// JSON keys to populate rows (per_day) or obtain final costing prices (finals)
			var keys = {};
				keys.finals = [];
				keys.per_day = [];
			var costing = {};
				costing.per_day = [];
				costing.finals = {};
			// table row data
			var days = $(tbody).find('tr');
			// the hidden input where the ACF field value will be stored
			var output = $costing_table_field.find('.acf-costing_table-output input');

			tfoot.find('tr').eq(1).find('th').each(function(i,th) {

				var $th = $(th);
				keys.finals.push($th.data('field'));

			});

			tfoot.find('tr').last().find('th input').each(function(i,input) {

				var $input = $(input);
				costing.finals[keys.finals[i]] = $input.val();

			});

			thead.find('tr').last().find('th').each(function(i,th) {

				var $th = $(th);
				if ( i > 0 ) {
					keys.per_day.push($th.data('field'));
				}

			});

			days.each(function(i,tr) {

				var obj = {};
				$(tr).find('td input').each(function(i,input) {

					var $input = $(input);
					obj[keys.per_day[i]] = $input.val();

				});

				costing.per_day.push(obj);

			});

			output.val(JSON.stringify(costing));
		}

		// calculate costings trigger on button click
		$( '.acf-costing_table-calc' ).on( 'click', function() {

			ACF_costing_table_calculate();
			ACF_costing_table_make_json();

		});

		// fire the json builder function also in the event an admin manually edits the totals inputs
		totals.find( 'input' ).on( 'change', function() {

			 ACF_costing_table_make_json();

		});

	}

	// ACF stuff
	if( typeof acf.add_action !== 'undefined' ) {

		/**
		 * Event triggered when page is loaded or new DOM elements are created by ACF
		 *
		 * @param	jQUery selection 	$el		the jQuery element which contains the ACF fields
		 *
		 * @since	1.0.0
		 */
		acf.add_action('ready append', function( $el ){

			// search $el for fields of type 'costing_table'
			acf.get_fields( { type : 'costing_table' }, $el ).each( function() {

				initialize_field( $(this) );

			});

		});

	}
	
})(jQuery);